### PR TITLE
Extrusion bounded

### PR DESCRIPTION
**Pull request by lynn-lumen**
# Objective

- Implement `Bounded3d` for some `Extrusion<T>`
- Provide methods to calculate `Aabb3d`s and `BoundingSphere`s for any extrusion with a `Bounded2d` base shape

## Solution

- Implemented `Bounded3d` for all 2D `bevy_math` primitives with the exception of `Plane2d`. As far as I can see, `Plane2d` is pretty much a line? and I think it is very unintuitive to extrude a plane and get a plane as a result.
- Add `extrusion_bounding_box` and `extrusion_bounding_sphere`. These are not always used internally since there are faster methods for specific extrusions. Both of them produce the optimal result within precision limits though.

## Testing

- Bounds for extrusions are tested within the same module. All unique implementations are tested.
- The correctness was validated visually aswell.

### CODE REVIEW

File: crates/bevy_math/src/bounding/bounded3d/extrusion.rs [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+        let top = segment_dir * self.half_depth;
+        let bottom = -top;
+
+        let e = Vec3::ONE - segment_dir * segment_dir;
+        let half_size = self.base_shape.radius * Vec3::new(e.x.sqrt(), e.y.sqrt(), e.z.sqrt());
+
+        Aabb3d {
+            min: (translation + (top - half_size).min(bottom - half_size)).into(),
+            max: (translation + (top + half_size).max(bottom + half_size)).into(),
+        }
```
[the remark this answers is not in the document]
Comment (IQuick143) [reply crates/bevy_math/src/bounding/bounded3d/extrusion.rs:30]: It could perhaps be called `depth` to be consistent with other implementations which use `Aabb3d::grow(depth.abs())`

**Review by IQuick143 — COMMENTED** [review]
Overall the code is good, although I haven't looked through the ellipse code and the tests yet.

File: crates/bevy_math/src/bounding/bounded3d/extrusion.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (solis-lumine-vorago) [reply crates/bevy_math/src/bounding/bounded3d/extrusion.rs]: Oh yeah, you are absolutely right :) That part was heavily inspired by the cylinder implementation 😅

File: crates/bevy_math/src/bounding/bounded3d/extrusion.rs [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+
+            if axis.element_product() == 0. {
+                return rotation * Vec3::new(a * axis.y, b * axis.x, 0.);
+            }
+            let m = -axis.x / axis.y;
+            let signum = axis.signum();
+
+            let y = signum.y * b * b / (b * b + m * m * a * a).sqrt();
+            let x = signum.x * a * (1. - y * y / b / b).sqrt();
+            rotation * Vec3::new(x, y, 0.)
```
Comment (NthTensor) [inline crates/bevy_math/src/bounding/bounded3d/extrusion.rs:59]: Possible division by zero if ellipse is degenerate?

**Review by NthTensor — APPROVED** [review]
Very cool! Everything looks as I would expect. Finding the Ellipse code a little hard to follow without comments, but think I get the gist.

Haven't visually inspected yet, will try to later. Approving.

**Review by alice-i-cecile — APPROVED** [review]

**Review by mweatherley — APPROVED** [review]
